llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -9607,7 +9624,7 @@ void VPReplicateRecipe::execute(VPTransformState &State) {
             VectorType::get(UI->getType(), State.VF));
         State.set(this, Poison);
       }
-      State.packScalarIntoVectorValue(this, *State.Lane);
+      State.packScalarIntoWideValue(this, *State.Lane);
```

Comment:
Given you're setting the state to potentially be a struct of vectors, what happens if we use the result in something that isn't extractvalue? For example, can you see what happens with IR like this?

```
  %call = tail call { float, float } @baz(float %in_val)
  store { float, float } %call, ptr %gep, align 8
```

? I'd hope that LoopVectorizationLegality prevents it, since you're only permitting return types to be structs. However, I do worry that essentially this patch is opening up vectorisation for more use cases than just extracting a value that we haven't tested yet. For the case of the store you would have to effectively treat this as an interleaving store to ensure the data is correctly laid out in memory.

In general though, these VPlan changes are essentially going down the path of defining what it means to vectorise any arbitrary operation that has a struct as either input or output. It sets in stone a single mechanism for vectorising such types, which is that a struct of scalars is treated as a struct of vectors once vectorised. However, I can also imagine cases where we might want to treat it as a vector of structs - see the store example above. @fhahn is this what you were worried about with this patch?

Reply:
As mentioned above, I've now updated loop LoopVectorizationLegality to only allow the case where the struct is produced by a call, and all users are extractvalue. I think this resolves the issue of dealing with structs in unsupported cases.

Reply:
I can't comment much on struct-of-vectors vs vector-of-structs, other than that this patch is very much focused on struct-of-vectors. The reason is this aligns with LLVM intrinsics that return multiple results and support vector types.

e.g. https://llvm.org/docs/LangRef.html#llvm-sincos-intrinsic,  https://llvm.org/docs/LangRef.html#llvm-sadd-with-overflow-intrinsics, or https://llvm.org/docs/LangRef.html#llvm-frexp-intrinsic

Reply:
Yep I think the potential issue @david-arm mentioned is similar to having vector functions that return multiple results as interleaved vector for example, where we lose some flexibility with the current schema in this patch.

Reply:
Is that a reason not to go ahead with this though? The basic struct vectorization added here is enough for some common use cases. I also don't think there's anything preventing future work if something more general is needed.